spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/RNN.scala

```diff
@@ -0,0 +1,389 @@
+
+        src_layer_MD = src_layer.getMemoryDescription()
+        src_iter_MD = src_iter.getMemoryDescription()
+        weights_layer_MD = wei_layer.getMemoryDescription()
+        weights_iter_MD = wei_iter.getMemoryDescription()
+        bis_MD = bis.getMemoryDescription()
+        dist_layer_MD = dst_layer.getMemoryDescription()
+        dist_iter_MD = dst_iter.getMemoryDescription()
+
+        /** TODO: user-defined initial hidden state is not supported currently.
+          * The default initial hidden state is all zero.
+          */
+        src_i = new TensorMMap(inputShape_iter)
+        src_i.dense.copy(Tensor[Float]().resize(inputShape_iter).zero())
+        dst_i = new TensorMMap(outputShape_iter)
+        dst_i.dense.copy(Tensor[Float]().resize(outputShape_iter).zero())
+
+      case Direction.BidirectionalSum =>
+        val weightShape = weight.size() /* ldigo */
+        val biasShape = bias.size() /* ldgo */
+        val outputShape = Array(inputShape(0), inputShape(1), hiddenSize) /* tnc */
+
+        val inputShape_iter = Array(common_n_layers, 2, nstates,
+          inputShape(1), hiddenSize) /* ldsnc */
+        val weightShape_iter = weight_i.size() /* ldigo */
+        val outputShape_iter = inputShape_iter /* ldsnc */
+
+        val src_layer = NativeData(inputShape, Memory.Format.any)
+        val src_iter = NativeData(inputShape_iter, Memory.Format.any)
+        val wei_layer = NativeData(weightShape, Memory.Format.any)
+        val wei_iter = NativeData(weightShape_iter, Memory.Format.any)
+        val bis = NativeData(biasShape, Memory.Format.any)
+        val dst_layer = NativeData(outputShape, Memory.Format.any)
+        val dst_iter = NativeData(outputShape_iter, Memory.Format.any)
+
+        src_layer_MD = src_layer.getMemoryDescription()
+        src_iter_MD = src_iter.getMemoryDescription()
+        weights_layer_MD = wei_layer.getMemoryDescription()
+        weights_iter_MD = wei_iter.getMemoryDescription()
+        bis_MD = bis.getMemoryDescription()
+        dist_layer_MD = dst_layer.getMemoryDescription()
+        dist_iter_MD = dst_iter.getMemoryDescription()
+
+        src_i = new TensorMMap(inputShape_iter)
+        src_i.dense.copy(Tensor[Float]().resize(inputShape_iter).zero())
+        dst_i = new TensorMMap(outputShape_iter)
+        dst_i.dense.copy(Tensor[Float]().resize(outputShape_iter).zero())
+
+      case _ => throw new UnsupportedOperationException("Not support such direction")
+    }
+  }
+
+  override private[mkldnn] def initFwdPrimitives(inputs: Array[MemoryData], phase: Phase) = {
+    val kind = if (!isTraining()) {
+      PropKind.ForwardInference
+    } else {
+      throw new UnsupportedOperationException("Not support training")
+    }
+
+    val rnnCellDesc = mode match {
+      case AlgKind.VanillaLstm =>
+        MklDnn.RNNCellDescInit(AlgKind.VanillaLstm, f, flags, alpha, clipping)
+      case _ => throw new UnsupportedOperationException("Not support such cell")
```

Comment:
print mode

Reply:
Several throwing exceptions have been updated to print more info. Input format check added. More members were added to release code.